jina/proto/message.py

```diff
@@ -0,0 +1,364 @@
+import os
+import sys
+from typing import List, Union, Optional
+
+from . import jina_pb2
+from .. import __version__, __proto_version__
+from ..enums import CompressAlgo
+from ..excepts import MismatchedVersion
+from ..logging import default_logger
+
+_trigger_body_fields = set(kk
+                           for v in [jina_pb2.Request.IndexRequest,
+                                     jina_pb2.Request.SearchRequest,
+                                     jina_pb2.Request.TrainRequest,
+                                     jina_pb2.Request.ControlRequest]
+                           for kk in v.DESCRIPTOR.fields_by_name.keys())
+
+_trigger_req_fields = set(jina_pb2.Request.DESCRIPTOR.fields_by_name.keys()).difference(
+    {'train', 'index', 'search', 'control'})
+
+_trigger_fields = _trigger_req_fields.union(_trigger_body_fields)
+_empty_request = jina_pb2.Request()
+
+__all__ = ['ProtoMessage', 'ControlMessage']
+
+
+class LazyRequest:
+    """
+    A container for serialized :class:`jina_pb2.Request` that only triggers deserialization
+    and decompression when receives the first read access to its member.
+
+    It overrides :meth:`__getattr__` to provide the same get/set interface as an
+    :class:`jina_pb2.Request` object.
+
+    """
+
+    def __init__(self, request: bytes, envelope: Optional['jina_pb2.Envelope'] = None):
+        self._buffer = request
+        self._deserialized = None  # type: jina_pb2.Request
+        self._envelope = envelope
+
+    @property
+    def is_used(self) -> bool:
+        """Return True when request has been r/w at least once """
+        return self._deserialized is not None
+
+    def __getattr__(self, name: str):
+        # https://docs.python.org/3/reference/datamodel.html#object.__getattr__
+        if (name in _trigger_fields) or hasattr(_empty_request, name):
+            self._deserialized = self.as_pb_object()
+        if name in _trigger_body_fields:
+            req = getattr(self._deserialized, self._deserialized.WhichOneof('body'))
+            return getattr(req, name)
+        elif hasattr(_empty_request, name):
```

Comment:
this branch will never be triggered. Since the same option is in an `or` statement in the first `if`

Reply:
it is triggered